src/viewport/viewport.service.ts

```diff
@@ -100,72 +52,50 @@ export class ViewportService {
 	 */
-	calculateItemsPerRow(containerWidth: number, itemWidth: number): number {
-		if (containerWidth === 0) {
-			return 0;
-		}
-
-		if (!containerWidth && !this.windowRef.hasNative) {
-			// todo: find a way to get container width for ssr
-			containerWidth = this.viewportServerSize.get().width;
-		}
-
-		return containerWidth / itemWidth;
-	}
-
 	private getViewportSize(): ViewportSize {
 		if (!this.windowRef.hasNative) {
 			return this.viewportServerSize.get();
 		}
 
 		const ua = navigator.userAgent.toLowerCase();
-		// safari subtracts the scrollbar width
-		if (ua.indexOf("safari") !== -1 && ua.indexOf("chrome") === -1) {
+		if (ua.indexOf("safari") !== -1 && ua.indexOf("chrome") === -1) { // safari subtracts the scrollbar width
 			return {
-				width: this.document.documentElement.clientWidth,
-				height: this.document.documentElement.clientHeight,
+				width: this.windowRef.native.document.documentElement.clientWidth,
+				height: this.windowRef.native.document.documentElement.clientHeight,
 			};
 		}
+
 		return {
 			width: this.windowRef.native.innerWidth,
 			height: this.windowRef.native.innerHeight,
 		};
 	}
 
-	private calculateViewportSize(width: number): ViewportSizeTypeInfo {
-		if (width === this.lastWidthCheck && this.lastWidthSizeInfo) {
-			return this.lastWidthSizeInfo;
-		}
-		this.lastWidthCheck = width;
+	/**
+	 * Returns the current viewport information
+	 * @param width the viewport width
+	 */
+	private getWidthSizeInfo(width: number): ViewportSizeTypeInfo {
+		const lastEntryIndex = this.viewportSizeTypeInfoList.length - 1;
 
-		// todo: make this more dynamic + decouple from lodash
-		if (_.inRange(width, viewportSizesConfig.xsmall)) {
-			this.lastWidthSizeInfo = viewportSizeRefs[ViewportSizeType.xsmall];
-		} else if (_.inRange(width, viewportSizesConfig.xsmall, viewportSizesConfig.small)) {
-			this.lastWidthSizeInfo = viewportSizeRefs[ViewportSizeType.small];
-		} else if (_.inRange(width, viewportSizesConfig.small, viewportSizesConfig.medium)) {
-			this.lastWidthSizeInfo = viewportSizeRefs[ViewportSizeType.medium];
-		} else if (_.inRange(width, viewportSizesConfig.medium, viewportSizesConfig.large)) {
-			this.lastWidthSizeInfo = viewportSizeRefs[ViewportSizeType.large];
-		} else if (_.inRange(width, viewportSizesConfig.large, viewportSizesConfig.xlarge)) {
-			this.lastWidthSizeInfo = viewportSizeRefs[ViewportSizeType.xlarge];
-		} else if (_.inRange(width, viewportSizesConfig.xlarge, viewportSizesConfig.xxlarge)) {
-			this.lastWidthSizeInfo = viewportSizeRefs[ViewportSizeType.xxlarge];
-		} else {
-			this.lastWidthSizeInfo = viewportSizeRefs[ViewportSizeType.xxlarge1];
+		for (let idx = 0; idx < lastEntryIndex; idx++) {
+			const viewportSizeTypeInfo = this.viewportSizeTypeInfoList[idx];
+
+			if (width <= viewportSizeTypeInfo.widthThreshold) {
+				return viewportSizeTypeInfo;
+			}
 		}
-		return this.lastWidthSizeInfo;
+
+		return this.viewportSizeTypeInfoList[lastEntryIndex];
```

Comment:
iterate through the ordered list while excluding the last 1. 

if not within the range, the last one is retrieved